src/wpool/mongoose_gun_worker.erl

```diff
@@ -0,0 +1,262 @@
+            Requests1 = maps:remove(StreamRef, Requests),
+            Requests2 = maps:merge(New, Requests1),
+
+            {noreply, State#state{requests = Requests2}}
+    end;
+%% `gun_down' is a message informing that Gun has lost the connection.
+%% It may try to reconnect, according to the `retry' and `retry_timeout' options,
+%% passed in the init options. They default to 5 retries, each with 5 second
+%% timeout.
+handle_info({gun_down, PID, Protocol, Reason, KilledStreams, UnprocessedStreams}, State) ->
+    ?WARNING_MSG("gun_down in mongoose_gun_worker. "
+                 "Gun has lost the ~p connection ~p because of \"~p\", killing ~p streams.",
+                 [Protocol, PID, Reason, length(KilledStreams) + length(UnprocessedStreams)]),
+    ?DEBUG("gun_down in mongoose_gun_worker. Killed streams: ~p. Unprocessed streams: ~p",
+           [KilledStreams, UnprocessedStreams]),
+    {noreply, State};
+%% `gun_up' is a message informing that Gun has reconnected after a `gun_down'.
+handle_info({gun_up, ConnPid, _Protocol},
+            State = #state{pid = ConnPid}) ->
+    ?DEBUG("gun_up in mongoose_gun_worker. Connection is back up with PID: ~p", [ConnPid]),
+    {noreply, State};
+%% `gun_error' is a message informing about any errors concerning connections or
+%% streams handled by Gun.
+handle_info({gun_error, ConnPid, Reason},
+            State = #state{pid = ConnPid}) ->
+    ?WARNING_MSG("gun_error in mongoose_gun_worker. Reason: ~p.", [Reason]),
+    {noreply, State};
+handle_info({gun_error, ConnPid, StreamRef, Reason},
+            State = #state{pid = ConnPid}) ->
+    ?WARNING_MSG("gun_error in mongoose_gun_worker. Reason: ~p. Stream reference: ~p",
+                 [Reason, StreamRef]),
+    {noreply, State};
+%% After `retry' number of `gun_down' messages, the connection process dies.
+%% Because it is monitored, we receive a `DOWN' message and may restart the
+%% connection.
+handle_info({'DOWN', MRef, process, ConnPid, Reason},
+            #state{pid = ConnPid, monitor = MRef} = State) ->
+    ?WARNING_MSG("Mongoose_gun_worker has lost the connection with PID ~p. Reason ~p.",
+                 [ConnPid, Reason]),
+    ConnectedState = open_connection(State),
+    NewState = retry_all(ConnectedState),
+    {noreply, NewState};
+handle_info(M, S) ->
+    ?ERROR_MSG("Unexpected message in gun_worker ~p", [M]),
+    {noreply, S}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+parse_uri(Host) ->
+    M = uri_string:parse(Host),
+    H = maps:get(host, M, "") ++ maps:get(path, M, ""),
+    Tls = case maps:get(scheme, M, undefined) of
+        "https" -> #{transport => tls};
+              _ -> #{}
+    end,
+    {H, maps:get(port, M, undefined), maps:merge(default_opts(), Tls)}.
+
+open_connection(State) ->
+    {ok, PID} = gun:open(State#state.host, State#state.port, State#state.opts),
```

Comment:
Ah, look at this: https://ninenines.eu/docs/en/gun/1.3/guide/connect/#_opening_a_new_connection
> If the port given is 443, Gun will attempt to connect using TLS. The protocol will be selected automatically using the ALPN extension for TLS. By default Gun supports HTTP/2 and HTTP/1.1 when connecting using TLS.
> For any other port, Gun will attempt to connect using plain TCP and will use the HTTP/1.1 protocol.

So it seems like we're not even using the http2 protocol at all. And actually, when tracing, I can see how gun is using http1 by default. So it has to probably be specified if it's possible, http2 doesn't necessarily require tls, so we need to decide how to match these two verticals (http-vs-http2 and tcp-vs-tls).

Reply:
Yeah, but the next sentence is 

> The transport and protocol used can be overriden via options. The manual documents all available options.

and when I was doing tests and printing out debug info, I have seen `http2` returned as the protocol from `gun:await_up/1`. If this changes during the tests to http1 than we have some very big problem, but I believe it is simply the case that most of the tests are not using http2 (it is not set as a default in `mongoose_gun_worker` and I believe nor is it anywhere else). In the `http_client_SUITE` there is a group devoted to http2.
The thing is that, given port 443, Gun will default to TLS. Because of that, it will then default to choosing http2 over http1. I believe that when leaving default options, this is the only time Gun will choose http2. But I don't see any warning on Gun's documentation stating that one cannot use http2 without TLS, or http2 on a different port. We could change the tests to use http2, but what would really be the purpose? This new pool should also work correctly with http1.